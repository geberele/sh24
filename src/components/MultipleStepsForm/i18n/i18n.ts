import { I18N } from './i18n.types'

export const en: I18N = {
  question1: 'What is your name?',
  question2: 'What is your email address?',
  question3: 'What service are you here for?',
  options1: 'STI Testing',
  options2: 'Contraception',
  options3: 'Other',
  labelName: 'Name',
  labelEmail: 'Email address',
  previewTitle: 'Preview',
  prev: 'Back',
  next: 'Next',
  errorMessage1: 'Please insert a valid name (2-25 characters)',
  errorMessage2: 'Please insert a valid email',
  errorMessage3: 'Please select an option',
}

export const it: I18N = {
  question1: 'Come ti chiami?',
  question2: 'Quale Ã¨ la tua email?',
  question3: 'Che tipo di servizio ti interessa?',
  options1: 'Test STI',
  options2: 'Contraccettivo',
  options3: 'Altro',
  labelName: 'Nome',
  labelEmail: 'Email',
  previewTitle: 'Preview',
  prev: 'Indietro',
  next: 'Avanti',
  errorMessage1: 'Inserisci un nome valido (2-25 lettere)',
  errorMessage2: 'Inserisci un\'email valida',
  errorMessage3: 'Seleziona un\'opzione per favore',
}
