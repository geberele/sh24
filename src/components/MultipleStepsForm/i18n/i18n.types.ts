export interface I18N {
  question1: string
  question2: string
  question3: string
  options1: string
  options2: string
  options3: string
  labelName: string
  labelEmail: string
  previewTitle: string
  prev: string
  next: string
  errorMessage1: string
  errorMessage2: string
  errorMessage3: string
}
