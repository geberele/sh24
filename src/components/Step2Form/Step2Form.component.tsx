import React from 'react'
import { useForm } from 'react-hook-form'

export interface Step2FormProps {
  label: string
  question: string
  next: string
  prev: string
  value: string
  errorMessage: string
  nextStep: () => void
  prevStep: () => void
  onChange: (value: string) => void
}

export const Step2Form = ({
  label,
  question,
  next,
  prev,
  value,
  nextStep,
  prevStep,
  onChange,
  errorMessage,
}: Step2FormProps) => {
  const { register, handleSubmit, errors } = useForm()

  return (
    <form onSubmit={handleSubmit(nextStep)}>
      <h1>{question}</h1>
      <label htmlFor="email">{label}</label>
      <input
        name="email"
        ref={register({
          required: true,
          pattern: {
            value: /^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
            message: errorMessage,
          },
        })}
        defaultValue={value}
        onChange={(e) => onChange(e.target.value)}
        data-testid="input-email"
      />

      {errors.email &&
      <div
        className="error"
        data-testid="input-error"
      >
        {errorMessage}
      </div>
      }

      <div className="navigation">
        <button
          onClick={prevStep}
          className="navigationLeft"
          data-testid="button-prev"
        >
          {prev}
        </button>
        <button
          type="submit"
          className="navigationRight"
          data-testid="button-next"
        >
          {next}
        </button>
      </div>
    </form>
  )
}
