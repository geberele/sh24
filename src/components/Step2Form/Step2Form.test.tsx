import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { fireEvent, waitFor, screen } from '@testing-library/dom'
import { Step2Form, Step2FormProps } from './Step2Form.component'
import { en } from '../MultipleStepsForm/i18n/i18n'

describe('Step2Form', () => {
  const prevStep = jest.fn()
  const nextStep = jest.fn()

  const props: Step2FormProps = {
    question: en.question2,
    label: en.labelEmail,
    prev: en.prev,
    next: en.next,
    value: name,
    prevStep: prevStep,
    nextStep: nextStep,
    onChange: () => {
      return
    },
    errorMessage: en.errorMessage2,
  }

  const setup = () => {
    const utils = render(<Step2Form {...props}/>)
    const input = screen.getByTestId('input-email') as HTMLInputElement
    const prevButton = screen.getByTestId('button-prev') as HTMLInputElement
    const submit = screen.getByTestId('button-next') as HTMLInputElement
    return {
      input,
      prevButton,
      submit,
      ...utils,
    }
  }

  it('renders', () => {
    const { asFragment } = setup()
    expect(asFragment()).toMatchSnapshot()
  })

  it('returns error if email is not valid', async () => {
    const { input, submit } = setup()
    fireEvent.change(input, { target: { value: 'test' } })
    fireEvent.click(submit)
    const error = await waitFor(() => screen.getByTestId('input-error'))
    expect(error.textContent).toBe(en.errorMessage2)
  })

  it('submits the form if email is valid', async () => {
    const { input, submit } = setup()
    fireEvent.change(input, { target: { value: 'test@example-mail.com' } })
    fireEvent.click(submit)
    await waitFor(() => expect(nextStep).toHaveBeenCalledTimes(1))
  })

  it('goes to the previous step', async () => {
    const { prevButton } = setup()
    fireEvent.click(prevButton)
    await waitFor(() => expect(prevStep).toHaveBeenCalledTimes(1))
  })
})
