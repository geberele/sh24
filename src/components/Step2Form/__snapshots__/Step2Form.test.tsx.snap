// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step2Form renders 1`] = `
<DocumentFragment>
  <form>
    <h1>
      What is your email address?
    </h1>
    <label
      for="email"
    >
      Email address
    </label>
    <input
      data-testid="input-email"
      name="email"
      value=""
    />
    <div
      class="navigation"
    >
      <button
        class="navigationLeft"
        data-testid="button-prev"
      >
        Back
      </button>
      <button
        class="navigationRight"
        data-testid="button-next"
        type="submit"
      >
        Next
      </button>
    </div>
  </form>
</DocumentFragment>
`;
